feat(Radio): added value, disabled, name, checked, defaultChecked and onChange props

mnajdova opened the pull request:
# Radio

This PR introduces the following props to the Radio component: value, disabled, checked, and onChange. 

### TODO

- [x] Conformance test
- [x] Minimal doc site example
- [ ] Stardust base theme
- [x] Teams Light theme
- [ ] Teams Dark theme
- [ ] Teams Contrast theme
- [x] Confirm RTL usage
- [ ] [W3 accessibility](https://www.w3.org/standards/webdesign/accessibility) check
- [ ] [Stardust accessibility](https://github.com/stardust-ui/accessibility) check
- [ ] Update glossary props table
- [x] Update the CHANGELOG.md

# API Proposal

## value

The HTML input value.

## disabled

A radio can appear disabled and be unable to change states.

```jsx
<Radio label="Disabled" disabled />
```

```html
<div class="ui-radio a b">
  <label class="ui-label j e f k l m i n o p d g h">
    <input type="radio" disabled="" class="c"/>Disabled
  </label>
</div>
```

## onChange

Callback invoked when the radio is clicked.

## checked
Whether or not the radio is checked.

```jsx
 <Radio label="This radio comes pre-checked" checked />
```

```html
<div class="ui-radio a b">
  <label class="ui-label j e f k l m i n o p d g h">
    <input type="radio" checked="" class="c"/>This radio comes pre-checked
  </label>
</div>
```

Commit:
-added value, disabled, checked, defaultChecked and onChange props to the Radio component

Commit:
-updated changelog

mnajdova commented on docs/src/examples/components/Radio/States/index.tsx:
```diff
@@ -0,0 +1,33 @@
+      examplePath="components/Radio/States/RadioExampleChecked"
+    >
+      <Segment style={{ marginTop: '10px' }}>
+        Use{' '}
```
I stole this from SUIR, I didn't know we can do this. Very cool! :D

levithomason committed:
Merge branch 'master' into feat/radio-checked

Commit:
-added name property on the Radio component

Commit:
Merge branch 'feat/radio-checked' of https://github.com/stardust-ui/react into feat/radio-checked

Commit:
-introduced RadioGroup component
-reverted the Radio to extend UIComponent

Commit:
-changed example in the radio group

Commit:
-small fix

Commit:
-removed auto controlled props

Commit:
-removed RadioGroup
-styles the Radio button with custom icon

Commit:
-adjusting the content of the Radio button and added variable for the icon's margin

kuzhelov commented on docs/src/examples/components/Radio/Types/RadioExample.shorthand.tsx:
```diff
@@ -1,6 +1,10 @@
 
-const RadioExample = () => <Radio label="Make your choice" />
+const handleChange = () => {
+  console.log('The radio checked value was changed!')
```
probably, it is better to use `alert` calls here, if the intent is to make it clear to the client that some changes have happened

mnajdova replied:
Makes sense, will change it!

kuzhelov commented on src/components/Button/ButtonGroup.tsx:
```diff
@@ -30,7 +30,7 @@ class ButtonGroup extends UIComponent<Extendable<IButtonGroupProps>, any> {
     as: customPropTypes.as,
 
-    /** A button can take the width of its container. */
+    /** The buttons contained inside the ButtonGroup. */
```
+1 👍

kuzhelov commented on src/components/Radio/Radio.tsx:
```diff
@@ -1,66 +1,157 @@
     className: PropTypes.string,
 
+    /** Initial checked value. */
+    defaultChecked: PropTypes.bool,
```
this one is not used anywhere - should we delete it?

kuzhelov replied:
also, it seems that this one is rather pertains to `RadioGroup` concept, rather than to individual `Radio`

mnajdova replied:
Every auto-controlled prop must have a default prop associated to it, that is why we have it. I may need to provide example usage of it.

mnajdova replied:
@kuzhelov , I added the usage of the default checked value. This value is intended to be used as a initial check value of the radio button. As the component is auto-controlled, the user should be able to initially set the value of the checked property. Please comment if you have some more doubts on this.

kuzhelov replied:
ok - but in that case let me share what my concerns are about. Now we have this `defaultChecked` to be defined in prop types, but there is no such a property in the component's TS interface - with that it, essentially, means that me as a client is not able to safely specify value for it in the TS code (I mean, this prop won't be advertised as one that is supported by the component, so I could provide some `string` value, for example).

mnajdova replied:
Sorry, that is my fault, the defaultChecked prop **SHOULD** be provide in the TS interface. Thanks for the catch!

kuzhelov commented on src/components/Radio/Radio.tsx:
```diff
@@ -1,66 +1,157 @@
+    }
+
+    if (onChange) {
+      onChange(e, { ...this.props, checked: !checked })
```
worrying about the fact that this callback method is invoked before state change will be applied to component - in case of potential error in the logic provided by the callback state change won't be applied

mnajdova replied:
Agreed, will move it before invoking the callback.

kuzhelov commented on src/components/Radio/Radio.tsx:
```diff
@@ -1,66 +1,157 @@
+
+  private handleChange = (e: React.SyntheticEvent) => {
+    const { onChange, disabled } = this.props
+    const { checked } = this.state
```
it would be more robust to obtain this flag directly from `e.target.checked` - due to async state updates there is a room (although very small one) for the `state` being out of sync with the component's visual look

mnajdova replied:
Makes sense. Will change it!

kuzhelov reviewed (approved):
couple of comments that we could discuss/address, other than that everything is 👍

Commit:
-removed the radio group example
-addressed comments in the PR

miroslavstastny commented on src/components/Radio/Radio.tsx:
```diff
@@ -1,66 +1,157 @@
+
+    this.trySetState({ checked })
+
+    if (onChange) {
```
`_.invoke()` does the check for you

mnajdova replied:
I replaced this with _.invoke, but I prefer much this type of invoking because I know for sure when clicking the property that it is used and invoke somewhere in the code, and with using the invoke, the name of the callback is just a string.

miroslavstastny commented on src/components/Radio/Radio.tsx:
```diff
@@ -1,66 +1,157 @@
+              circular: true,
+              size: 'mini',
+              variables: {
+                color: checked ? variables.checkedIconColor : variables.uncheckedIconColor,
```
I see you are already discussing this in #207. It would be great if, eventually, we would not need to touch variables in component code.

kuzhelov replied:
unfortunately, this makes 'variables defined in theme' story a bit more complicated, as now we might need to get an access to component props there - as it is discussed in #207. This would be quite significant move, and we should consider whether there are any other not-so-radical alternatives. Still, totally agree that we should finally come up with solution that will eliminate this code from component's logic

mnajdova replied:
Exactly, I was not satisfied with having to use the variable in the renderComponent, but as I mention in the other PR we should either be able to define the variables depending on the state and props, or rethink about the API provided so far for the components.

Commit:
-added defaultChecked in the TS interface
-replaced onChange with _.invoke

Commit:
Merge branch 'master' into feat/radio-checked

# Conflicts:
#	CHANGELOG.md

Commit:
-change the color of the label when the radio is disabled

Commit:
-renamed the test for the radio component
-changed conformance test regarding the transparent handling of the events to check that the second argument is object containing all props instead of just all props
-changed checked value to be read from state instead of event

Commit:
Merge branch 'master' into feat/radio-checked

Commit:
Merge branch 'master' into feat/radio-checked

Commit:
-added icon slot in the radio's variables

Commit:
-added icon slot in the radio's props interface
-fixed console error for onChange not being provided to the input

Commit:
-fixed the test in the radio
